### PR TITLE
[SYNCOPE-1635] Use singleton bean for each domain for each rule with conf

### PR DESCRIPTION
**Pull request by DmitriyBrashevets**
…conf

### CODE REVIEW

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java [hunk 1]
```diff
@@ -72,7 +73,7 @@
         }
     }
 
-    public static Optional<AccountRule> buildAccountRule(final Implementation impl)
+    public static Optional<AccountRule> buildAccountRule(final String domain, final Implementation impl)
```
Comment (ilgrosso) [inline core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:76]: same as above

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java [hunk 2]
```diff
@@ -96,7 +97,7 @@
         }
     }
 
-    public static Optional<PasswordRule> buildPasswordRule(final Implementation impl)
+    public static Optional<PasswordRule> buildPasswordRule(final String domain, final Implementation impl)
```
Comment (ilgrosso) [inline core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:100]: same as above

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java [hunk 3]
```diff
@@ -120,7 +121,7 @@
         }
     }
 
-    public static Optional<PullCorrelationRule> buildPullCorrelationRule(final Implementation impl)
+    public static Optional<PullCorrelationRule> buildPullCorrelationRule(final String domain, final Implementation impl)
```
Comment (ilgrosso) [inline core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:124]: same as above

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java [hunk 4]
```diff
@@ -145,7 +146,7 @@
         }
     }
 
-    public static Optional<PushCorrelationRule> buildPushCorrelationRule(final Implementation impl)
+    public static Optional<PushCorrelationRule> buildPushCorrelationRule(final String domain, final Implementation impl)
```
Comment (ilgrosso) [inline core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:149]: same as above

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java [hunk 5]
```diff
@@ -216,28 +217,27 @@
...
-                try {
-                    bean = (T) ApplicationContextProvider.getBeanFactory().
-                            createBean(clazz, AbstractBeanDefinition.AUTOWIRE_BY_TYPE, false);
-                    ApplicationContextProvider.getBeanFactory().registerSingleton(clazz.getName(), bean);
-                } catch (IllegalStateException e) {
-                    LOG.debug("While attempting to register {}", clazz.getName(), e);
+            String domainableBeanNameWithConf = domain + clazz.getName();
+            DefaultListableBeanFactory beanFactory = ApplicationContextProvider.getBeanFactory();
+
+            if (beanFactory.containsSingleton(domainableBeanNameWithConf)) {
```
Comment (ilgrosso) [inline core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:225]: The same exact `if` with statement is also performed in the `synchronized` block: why?

Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:225]: This is a double-check lock pattern. Thread takes monitor and executes the logic regarding initialization. Another thread waits until the the monitor is realeased. As soon as first thread releases monitor -> second threads proceeds its work and reaches the additional `if` clause. As the bean already exists (created by first thread) the second thread is not creating a bean again and not registers it as a singleton.


Do you see any disadvantages it this?

Comment (ilgrosso) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:225]: I see, there is some literature abut double-check lock pattern, and some is not very pleasant with it, especially for Java implementations.

What would be the disadvantage of simply removing the first `if` and leaving the whole logic in the `synchronized` block?

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java [hunk 6]
```diff
@@ -48,7 +49,7 @@
 
     private static final Map<String, Class<?>> CLASS_CACHE = Collections.synchronizedMap(new HashMap<>());
 
-    public static Optional<Reportlet> buildReportlet(final Implementation impl)
+    public static Optional<Reportlet> buildReportlet(final String domain, final Implementation impl)
```
[the remark this answers is not in the document]
Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:52]: I thought that it's becoming more understandable (visible) if public methods have the additional parameter declared explicitly. But can be removed...

[the remark this answers is not in the document]
Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java]: removed addtional param

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java (hunk 1 above)
Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java]: removed addtional param

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java (hunk 2 above)
Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java]: removed addtional param

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java (hunk 3 above)
Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java]: removed addtional param

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java (hunk 4 above)
Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java]: removed addtional param

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java (hunk 5 above)
Comment (DmitriyBrashevets) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:226]: Ah... I see... Maybe java optimization will remove this logic from bytecode when the method will become hot.

**Review by ilgrosso — APPROVED** [review]

File: core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java (hunk 5 above)
Comment (ilgrosso) [reply core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java:226]: I am not completely convinced, but let that be :-)

**Pull request merged**